lib/src/default_index/revset_engine.rs

```diff
@@ -1049,19 +1049,39 @@ fn build_predicate_fn(
                 pattern.matches(commit.description())
             })
         }
-        RevsetFilterPredicate::Author(pattern) => {
+        RevsetFilterPredicate::Author(pattern, mailmap) => {
             let pattern = pattern.clone();
+            let mailmap = mailmap.clone();
             // TODO: Make these functions that take a needle to search for accept some
             // syntax for specifying whether it's a regex and whether it's
             // case-sensitive.
+            box_pure_predicate_fn(move |index, pos| {
+                let entry = index.entry_by_pos(pos);
+                let commit = store.get_commit(&entry.commit_id()).unwrap();
+                let author = mailmap.author(&commit);
+                pattern.matches(&author.name) || pattern.matches(&author.email)
+            })
```

Comment:
Can't we transform the query by frontend?

```
author(foo)
-> author(foo)|author(exact:lookup_reverse(foo)[0])|author(exact:lookup_reverse(foo)[1])|..
```

It's not exactly the same (because the query also matches the original values), but I think it's practically okay. I think it's easier for revset engine backed by database to process query.

Reply:
Right, I mentioned this in the commit message under “The mapping to canonical signatures is one‐way, and queries only match on the canonical form”. It seems a little tricky to translate back because the functions accept glob patterns; at the very least we would have to traverse the `Mailmap` structure and syntheize `author_raw(exact:…)` patterns for all possible matches. But I’m not entirely sure that would be correct in all cases, even if we accept the change in semantics. I’d have to think about it.

Rather than changing the semantics/implementation here, I think in a database backend, you could have a caching table like (commit ID or hash or something of `.mailmap`, raw signature that appears in commits, canonical signature from corresponding `.mailmap`), and an index for lookup? I don’t really know how those backends work since there isn’t a public example of one. Maybe you could even use a materialized view or something.

Reply:
I don't know the implementation either, and I guess Google backend can just ignore mailmap thingy at all (because .mailmap shouldn't exist there.)

That said, it'll be nice if mailmap handling is isolated to the frontend layer. Suppose the size of the mailmap is small compared to the commit history, I think it's okay to scan all entries to build a reverse mapping.

Reply:
I had a think about this. If you wanted to avoid the additional table/view I mentioned, we can still obtain the same database performance benefits without changing semantics (which I’m a little reluctant to do, as it goes against the rubric of when raw signature content should affect UI I put in the commit message). `author(a*b*)` would optimize to `author(a*b*) & (author_raw(a*b*) | author_raw(exact:…) | author_raw(exact:…) | …)` based on the `.mailmap`. The latter can be queried on the database efficiently, and the `author(a*b*) &` is a quick post‐processing step on the results that ensures the semantics remain the same.

This wouldn’t be hard to implement, but I’m hesitant to add the complexity off the bat if it wouldn’t currently have any prospective users, especially as it would make performance strictly worse with the current backends. (Although considering the complaints I’ve heard from Googlers in the past about what a pain it is to get their username and name updated across all systems maybe they could do with having some kind of system here…)

Reply:
I don't think doing preprocessing is strictly worse in performance. If the query doesn't match any mailmap entry (which I think is common), there's no additional cost in later stage. OTOH, without preprocessing, we'll do binary search for each commit entry. I don't think the difference would be significant, though.

FWIW, if a database-backed revset engine transforms mailmap to table/view, they'll want a single mailmap resource than mailmap per expression node.

Anyway, I think this is a fundamental issue of mailmap design, and I just thought it would be nice if we can avoid spreading a mess to the backend layer.

Reply:
The reason I say it would be strictly slower with the Git and local backends is that it would be increasing the number of comparisons. However, this probably isn’t really relevant. I take it that database backends don’t use `jj_lib::default_index::revset_engine` at all, then?

I’d like to discuss my semantics‐preserving solution further, if that’s alright. I think that desugaring/optimizing `author("new nam*")` to `author("new nam*") & (author_raw("new nam*") | author_raw(exact:"old name") | author_raw(exact:"old@email.example"))` would hopefully satisfy everyone here: it would provide all the same query performance advantages, not require any database schema adjustments, and be a very small burden on any implementors of alternate revset engines that want to support `.mailmap`. The idea is that there would be indexes on the raw signatures (as there are currently, presumably), and that the database’s optimized pattern query functions could be used on those, with the results then being post‐processed in either Rust code or the query language to ensure that they match the mapped name.

For instance, an index‐friendly query like the following could be issued, just like currently…

```sql
select …
from commits
where author_raw_name like "new nam%"
or author_raw_email like "new nam%"
or author_raw_name = "old name"
or author_raw_email = "old name"
or author_raw_name = "old@email.example"
or author_raw_email = "old@email.example"
```

…followed by a simple and quick `StringPattern::parse("new nam*").matches(mailmap.author(&commit))` filtering step in Rust to handle the `author("new nam*") &` condition. Obviously this is a common pattern when issuing database queries; I don’t know if every single revset function is translated to a direct database query filter with no Rust code currently, but if it’s desired for that property to be preserved then the post‐processing could also be implemented in the database query; any planner worth its name would do the fast index lookups before checking the additional condition, as with other queries that use complex functions. If there are already filter steps using Rust code, on the other hand, then this should “just work” with no additional logic or changes.

Anyway, I’m happy to implement this reverse‐lookup desugaring as requested, but if it’d slow things down on the in‐tree backends and Google doesn’t plan to use it, is it worth the additional logic right now? Maybe it’s just enough to know that we could do this without immediately implementing it?

Reply:
Maybe it would be a good idea to make backends responsible for providing a `Mailmap` (which could be a trait instead?), and the Git and local backends could just share the current implementation for now. Again though I’m unsure how much value this would add if it wouldn’t matter to Google right now anyway.

Reply:
Strawman design:

* `Mailmap` could become a trait.

* The `Backend` trait could gain e.g. `fn read_mailmap(&self, as_of_commit: &CommitId) -> impl Mailmap`. A database backend could stub this out if it doesn’t support the feature or ignore the parameter and return an interface to a database table. Handling the queries efficiently would be the responsibility of a backend that wants to support this, and could be handled as per the desugaring I outlined in https://github.com/martinvonz/jj/pull/3951#discussion_r1654369716.

* The in‐tree backends would share the current implementation.

Not sure if this is better or worse than the status quo, but I thought I should throw it out there.

Reply:
> Maybe it would be a good idea to make backends responsible for providing a `Mailmap` (which could be a trait instead?), and the Git and local backends could just share the current implementation for now.

Yeah, I think that's proper design for identity rewriting.

My understanding is that mailmap is ad-hoc solution and we decided to process mailmap at the frontend layer because of that? And, my point is that it's nice if mailmap handling does not cross the frontend-backend boudary. It might be a bit slow and inaccurate depending on the query, but I assume it's practically okay.

Reply:
> And, my point is that it's nice if mailmap handling does not cross the frontend-backend boudary.

I agree with this. At the same time, mailmap is a bit of a hack, so what Emily has currently implemented in this PR seems fine to me.

Another alternative might be to pass the `Mailmap` into `Index::evaluate_revset()`. What do you think about that?

Reply:
> Another alternative might be to pass the `Mailmap` into `Index::evaluate_revset()`.

That seems slightly better.

FWIW, I also consider mailmap as a hack, and that's the reason why I insist that the query result doesn't have to be 100% accurate. It's simpler to approximate mailmap query as union of matched mailmap entries than extending backend API.

Reply:
> Strawman design:
> 
> * `Mailmap` could become a trait.
> * The `Backend` trait could gain e.g. `fn read_mailmap(&self, as_of_commit: &CommitId) -> impl Mailmap`. A database backend could stub this out if it doesn’t support the feature or ignore the parameter and return an interface to a database table. Handling the queries efficiently would be the responsibility of a backend that wants to support this, and could be handled as per the desugaring I outlined in [lib: add support for `.mailmap` files #3951 (comment)](https://github.com/martinvonz/jj/pull/3951#discussion_r1654369716).
> * The in‐tree backends would share the current implementation.
> 
> Not sure if this is better or worse than the status quo, but I thought I should throw it out there.

Yes, that'd be the approach if the pursue the identity work further, but clearly is unneeded complexity for this PR. As what the current `.mailmap` approach should be, there I heavily agree with Yuya and Martin.

Reply:
> Another alternative might be to pass the `Mailmap` into `Index::evaluate_revset()`. What do you think about that?

I tried this, and it was a mess, sadly; the state needed to be piped through the whole evaluation system and it caused a lot of churn and other problems. It basically resulted in `ResolvedExpression`s not being fully resolved and having to patch that up in the evaluation layer, so in terms of these layering concerns it was a lot worse.

I wasn’t sure about how to move forward here, but then I realized that it is possible to implement the front‐end desugaring of the revset queries in a way that always gives correct `.mailmap` results and yields simple, efficient `RevsetExpression`s that should work great with database indexes. The backend now no longer needs to care about `.mailmap` support at all. Hopefully that will make everyone happy! I’m still finishing polishing it off, so I haven’t pushed it here yet, but I’ve opened https://github.com/martinvonz/jj/pull/3991 as an interim change I made along the way that resolves an old TODO and helped with the implementation.

(For what it’s worth: I do agree that `.mailmap` is essentially a pragmatic hack, an inherently incomplete measure, and a symptom of a deficient underlying design. Still, Git users are unlikely to ever get anything better, and for some people it’s a hack that is the difference between a marginally workable identity system and one that is completely unfit for purpose, so I do care about implementing it at least as well as Git does. I’m principled in my pragmatism; giving up on the principles of application I listed in the commit message results in concrete scenarios where it fails to meet the basic point of supporting `.mailmap`, in my view. Thankfully, it looks like it’s totally practical to address these layering concerns while still upholding them!)